src/main/scala/cognite/spark/v1/NumericDataPointsRdd.scala

```diff
@@ -422,54 +434,81 @@ case class NumericDataPointsRdd(
             id,
             dataPointsRange.start,
             dataPointsRange.end)
+          .map { allDataPoints =>
+            increaseReadMetrics(allDataPoints.length)
+            allDataPoints.map(p => toRowDataPointsRange(dataPointsRange.id, p))
+          }
+    }
+    Stream.evalUnChunk(points.map(Chunk.seq))
+  }
+
+  @inline
+  private def toRowDataPointsRange(id: Either[Long, String], dataPoint: SdkDataPoint): Row = {
+    val array = new Array[Any](rowIndices.length)
+    var i = 0
+    for (f <- rowIndices) {
+      f match {
+        case 0 => array(i) = id.left.toOption
+        case 1 => array(i) = id.right.toOption
+        case 2 => array(i) = java.sql.Timestamp.from(dataPoint.timestamp)
+        case 3 => array(i) = dataPoint.value
+        case 4 | 5 => array(i) = None
+      }
+      i += 1
+    }
+    new GenericRow(array)
+  }
+
+  @inline
+  private def toRowAggregationRange(r: AggregationRange, dataPoint: SdkDataPoint): Row = {
+    val array = new Array[Any](rowIndices.length)
+    var i = 0
+    for (f <- rowIndices) {
+      f match {
+        case 0 => array(i) = r.id.left.toOption
+        case 1 => array(i) = r.id.right.toOption
+        case 2 => array(i) = java.sql.Timestamp.from(dataPoint.timestamp)
+        case 3 => array(i) = dataPoint.value
+        case 4 => array(i) = Some(r.aggregation)
+        case 5 => array(i) = Some(r.granularity.toString)
+      }
+      i += 1
+    }
+    new GenericRow(array)
+  }
+
+  private def maybeLimitStream[A](stream: Stream[IO, A]) =
+    config.limitPerPartition match {
+      case Some(limit) => stream.take(limit)
+      case None => stream
     }
 
   override def compute(_split: Partition, context: TaskContext): Iterator[Row] = {
     val bucket = _split.asInstanceOf[Bucket]
 
-    bucket.ranges.toVector
-      .parFlatTraverse {
-        case r: DataPointsRange =>
-          queryDataPointsRange(r)
-            .map(
-              dataPoints =>
-                dataPoints
-                  .map { p =>
-                    DataPointsItem(
-                      r.id.left.toOption,
-                      r.id.right.toOption,
-                      p.timestamp,
-                      p.value,
-                      None,
-                      None)
-                  }
-                  .map(toRow)
-                  .toVector)
+    val stream: Stream[IO, Stream[IO, Row]] = Stream
+      .emits(bucket.ranges.toVector)
+      .covary[IO]
+      .parEvalMapUnordered(100) {
```

Comment:
Where does this constant comes from? Should it be dynamic?

Reply:
We really just want it to be limited only by thread pools etc., turns out parJoinUnbounded sets this to Int.MaxValue and parEvalMapUnordered is using parJoin. We should use Int.MaxValue here as well, and I've added comments to explain this.